### ISSUE TITLE
`String` `Debug` implementation prints escaped single quotes

### ISSUE THREAD

**Issue by marcianx** [rust-lang/rust#83046]
This is very minor.

```rust
println!("{:?}", "'");
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f7440f769e43a29fb7ec66ec62225d1d))

Expected output: `"'"`
Actual output: `"\'"`

Perhaps this was intentional, but I'm not sure why it would be. The closest issue I saw was https://github.com/rust-lang/rust/issues/3752 which is ancient and was more obviously a bug since escapes were appearing in `Display` output and not just `Debug`.

This reproduces in the latest stable (1.50.0) and nightly (1.52.0-nightly) builds.

**Comment by osa1** [comment]
This is because the same format string method of `char` (`escape_debug_ext`) is used for `Debug` implementation of both `str` and `char` (and a few other internal types), so it needs to escake `'` otherwise in char context it would generate incorrect `Debug` string. Adding a `is_char` argument to this method solves the issue. I tested this and it passes ui tests.

I checked the language spec https://doc.rust-lang.org/stable/reference/tokens.html#string-literals I don't think `"\'"` is a valid string, right? Am I missing anything?
